src/Payments.sol

```diff
@@ -0,0 +1,572 @@
+
+        // --- Process the One-Time Payment ---
+        processOneTimePayment(payer, payee, oneTimePayment);
+
+        require(rail.lockupFixed >= 0, "rail lockupFixed must be non-negative");
+        require(payer.lockupCurrent >= 0, "payer lockupCurrent must be non-negative");
+        require(payer.lockupCurrent <= payer.funds, "payer lockup cannot exceed funds");
+
+        return 0;
+    }
+
+    function processOneTimePayment(
+        Account storage payer,
+        Account storage payee,
+        uint256 oneTimePayment
+    ) internal {
+        if (oneTimePayment > 0) {
+            require(payer.funds >= oneTimePayment, "insufficient funds for one-time payment");
+            payer.funds -= oneTimePayment;
+            payee.funds += oneTimePayment;
+        }
+    }
+
+    function validateAndModifyRateChangeApproval(
+        Rail storage rail,
+        OperatorApproval storage approval,
+        uint256 oldRate,
+        uint256 newRate,
+        uint256 oneTimePayment
+    ) internal {
+        require(oneTimePayment <= rail.lockupFixed, "one-time payment exceeds rail fixed lockup");
+
+        uint256 oldLockup = (oldRate * rail.lockupPeriod) + rail.lockupFixed;
+        uint256 newLockup = (newRate * rail.lockupPeriod) + (rail.lockupFixed - oneTimePayment);
+
+        // Check that new total lockup + one time payment is within allowance
+        require(newLockup + oneTimePayment <= oldLockup + approval.lockupAllowance, "exceeds operator lockup allowance");
+        // Adjust the operator's available lockup allowance.
+        approval.lockupAllowance = approval.lockupAllowance + oldLockup - (newLockup + oneTimePayment);
+        require(newRate <= approval.rateAllowance, "new rate exceeds operator rate allowance");
+    }
+
+
+    function updateRailArbiter(uint256 railId, address newArbiter) external validateRailActive(railId) onlyRailOperator(railId) {
+        Rail storage rail = rails[railId];
+        OperatorApproval storage approval = operatorApprovals[rail.token][rail.from][rail.operator];
+        require(approval.isApproved, "operator not approved");
+        if (approval.arbitrer != address(0)) {
+            require(newArbiter == approval.arbitrer, "arbiter mismatch");
+        }
+        // Update the arbiter
+        rail.arbiter = newArbiter;
+    }
+
+
+    function settleRail(uint256 railId, uint256 untilEpoch)
+        public
+        validateRailActive(railId)
+        validateRailAccountsExist(railId)
+        returns (
+            uint256 totalSettledAmount,
+            uint256 finalSettledEpoch
+        )
+    {
+        // Get the rail and the involved accounts.
+        Rail storage rail = rails[railId];
+        Account storage payer = accounts[rail.token][rail.from];
+        Account storage payee = accounts[rail.token][rail.to];
+
+        settleAccountLockup(payer);
+
+        uint256 settlementTargetEpoch = min(untilEpoch, payer.lockupLastSettledAt + rail.lockupPeriod);
```

Comment:
Need to check that `untilEpoch` is at or before the current epoch.

Reply:
Done.